README.md

# scReClassify <img src='img/scReClassify_sticker.png' align="right" height="138.5" />

`scReClassify` is a post hoc cell type classification of single-cell
RNA-sequencing data. Using semi-supervised learning algorithm,
adaSampling, to correct cell type annotation from noise.

# Getting started

## Vignette

For more detailed instuction, you can find the vignette at our website:
<https://sydneybiox.github.io/scdney/>.

### Installation

#### Bioconductor

```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("scReClassify")
```

#### GitHub

``` r
devtools::install_github("SydneyBioX/scReClassify", build_opts = c("--no-resave-data", "--no-manual"))
library(scReClassify)
```

For `devtool version
< 2.0.0`,

``` r
devtools::install_github("SydneyBioX/scReClassify", build_vignettes = TRUE)
library(scReClassify)
```


``` r
suppressPackageStartupMessages({
    library(scReClassify)
    library(SingleCellExperiment)
    library(SummarizedExperiment)
})
```


# Usage

![alt text](https://github.com/SydneyBioX/scReClassify/raw/master/img/scReClassify.jpg)


Current version of this package is implemented to run with `svm` and `randomForest` classifiers.

## Load data

``` r
data(gse87795_subset_sce)
dat <- gse87795_subset_sce
cellTypes <- gse87795_subset_sce$cellTypes

# number of clusters
nCs <- length(table(cellTypes))

# This demo dataset is already pre-processed
dim(dat)
```

## Part A. scReClassify (Demonstration)

### Dimension reduction

``` r
reducedDim(dat, "matPCs") = matPCs(dat, assay = "logNorm", 0.7)
```

### Synthetic noise (Demonstration purpose)

Here in this example, we will synthetically generate varying degree of noise in sample labels.

``` r
lab <- cellTypes

set.seed(1)
noisyCls <- function(dat, rho, cls.truth){
    cls.noisy <- cls.truth
    names(cls.noisy) <- colnames(dat)
    for(i in seq_len(length(table(cls.noisy)))) {
        # class label starts from 0
        if (i != length(table(cls.noisy))) {
            cls.noisy[sample(which(cls.truth == names(table(cls.noisy))[i]), floor(sum(cls.truth == names(table(cls.noisy))[i]) * rho))] <- names(table(cls.noisy))[i+1]
        } else {
            cls.noisy[sample(which(cls.truth == names(table(cls.noisy))[i]), floor(sum(cls.truth == names(table(cls.noisy))[i]) * rho))] <- names(table(cls.noisy))[1]
        }
    }
  
    print(sum(cls.truth != cls.noisy))
    return(cls.noisy)
}

cls.noisy01 <- noisyCls(t(reducedDim(dat, "matPCs")), rho=0.1, lab)
cls.noisy02 <- noisyCls(t(reducedDim(dat, "matPCs")), rho=0.2, lab)
cls.noisy03 <- noisyCls(t(reducedDim(dat, "matPCs")), rho=0.3, lab)
cls.noisy04 <- noisyCls(t(reducedDim(dat, "matPCs")), rho=0.4, lab)
cls.noisy05 <- noisyCls(t(reducedDim(dat, "matPCs")), rho=0.5, lab)
```

### Use scReClassify to correct mislabeled cell types.

Here in this example, we will only use `Support Vector machine (svm)` as base classifier.

#### Benchmark evaluation

``` r
###################################
# SVM
###################################
base <- "svm"
set.seed(1)
result = lapply(seq_len(10), function(j) {
    final <- multiAdaSampling(dat, cls.noisy01, reducedDimName = "matPCs", 
                                classifier=base, percent=1, L=10)$final
    ari01 <- mclust::adjustedRandIndex(lab, final)
    acc01 <- bAccuracy(lab, final)
    
    final <- multiAdaSampling(dat, cls.noisy02, reducedDimName = "matPCs", 
                                classifier=base, percent=1, L=10)$final
    ari02 <- mclust::adjustedRandIndex(lab, final)
    acc02 <- bAccuracy(lab, final)
    
    final <- multiAdaSampling(dat, cls.noisy03, reducedDimName = "matPCs", 
                                classifier=base, percent=1, L=10)$final
    ari03 <- mclust::adjustedRandIndex(lab, final)
    acc03 <- bAccuracy(lab, final)
    
    final <- multiAdaSampling(dat, cls.noisy04, reducedDimName = "matPCs", 
                                classifier=base, percent=1, L=10)$final
    ari04 <- mclust::adjustedRandIndex(lab, final)
    acc04 <- bAccuracy(lab, final)
    
    final <- multiAdaSampling(dat, cls.noisy05, reducedDimName = "matPCs", 
                                classifier=base, percent=1, L=10)$final
    ari05 <- mclust::adjustedRandIndex(lab, final)
    acc05 <- bAccuracy(lab, final)
    
    c(
      acc01 = acc01,
      acc02 = acc02,
      acc03 = acc03,
      acc04 = acc04,
      acc05 = acc05,
      ari01 = ari01,
      ari02 = ari02,
      ari03 = ari03,
      ari04 = ari04,
      ari05 = ari05
    )
})

result = do.call(rbind, result)
acc = result[,seq_len(5)]
colnames(acc) = seq(from=0.1,to=0.5,by=0.1)

ari = result[,seq(from= 6, to = 10)]
colnames(ari) = seq(from=0.1,to=0.5,by=0.1)


plot.new()
par(mfrow = c(1,2))
boxplot(acc, col="lightblue", main="SVM Accuracy", 
        ylim=c(0.45, 1), xlab = "rho", ylab = "Accuracy")
points(x=seq_len(5), y=c(
    bAccuracy(lab, cls.noisy01), 
    bAccuracy(lab, cls.noisy02),
    bAccuracy(lab, cls.noisy03), 
    bAccuracy(lab, cls.noisy04),
    bAccuracy(lab, cls.noisy05)), 
    col="red3", pch=c(2,3,4,5,6), cex=1)
boxplot(ari, col="lightblue", main="SVM ARI", 
        ylim=c(0.25, 1), xlab = "rho", ylab = "ARI")
points(x=seq_len(5), y=c(
    mclust::adjustedRandIndex(lab, cls.noisy01), 
    mclust::adjustedRandIndex(lab, cls.noisy02),
    mclust::adjustedRandIndex(lab, cls.noisy03), 
    mclust::adjustedRandIndex(lab, cls.noisy04),
    mclust::adjustedRandIndex(lab, cls.noisy05)), 
    col="red3", pch=c(2,3,4,5,6), cex=1)
```


## Part B. scReClassify (mislabeled cell type correction)

``` r
# PCA procedure
reducedDim(dat, "matPCs") = matPCs(dat, assay = "logNorm", 0.7)


# run scReClassify
set.seed(1)
cellTypes.reclassify <- multiAdaSampling(dat, cellTypes, 
                                        reducedDimName = "matPCs",
                                        classifier = "svm", percent = 1, L = 10)

# Verification by marker genes
End <- c("ITGA2B", "ITGB3")

# check examples
idx <- which(cellTypes.reclassify$final != cellTypes)

cbind(original=cellTypes[idx], reclassify=cellTypes.reclassify$final[idx]) %>%
    DT::datatable()

mat <- assay(dat, "logNorm")

c1 <- mat[, which(cellTypes=="Endothelial Cell")]
c2 <- mat[, which(cellTypes=="Erythrocyte")]
c3 <- mat[, which(cellTypes=="Hepatoblast")]
c4 <- mat[, which(cellTypes=="Macrophage")]
c5 <- mat[, which(cellTypes=="Megakaryocyte")]
c6 <- mat[, which(cellTypes=="Mesenchymal Cell")]
cs <- rainbow(length(table(cellTypes)))


# (example 1 E13.5_C14)
#####
par(mfrow=c(1,2))
marker <- End[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], 
        c4[marker,], c5[marker,], c6[marker,], 
        col=cs, main=marker, 
        names=c("Others", "Others", "Others", "Orignal", 
                "Reclassified", "Others"), las=2, xlab = "Labels",
        ylab = "log2FPKM")
points(5, mat[marker, which(colnames(mat) %in% "E13.5_C14")], 
        pch=16, col="red", cex=2)

marker <- End[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], 
        c4[marker,], c5[marker,], c6[marker,], 
        col=cs, main=marker, 
        names=c("Others", "Others", "Others", "Orignal", 
                "Reclassified", "Others"), las=2, xlab = "Labels", 
        ylab = "log2FPKM")
points(5, mat[marker, which(colnames(mat) %in% "E13.5_C14")], 
        pch=16, col="red", cex=2)
```


# Reference
scReClassify is published in BMC Genomics. Please refer to the following article for more details on method implementation and evaluation.

Taiyun Kim, Kitty Lo, Thomas A. Geddes, Hani Jieun Kim, Jean Yee Hwa Yang & Pengyi Yang (2019) scReClassify: post hoc cell type classification of single-cell RNA-seq data. BMC Genomics, 20:913. https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6305-x


NEWS.md
## scReCkassify 0.99.8
* Included citation for the package

## scReCkassify 0.99.7
* Removed global assignment in multiAdaSampling.

## scReCkassify 0.99.6
* Revised further to address comments/feedbacks

## scReCkassify 0.99.5
* Fixed NEWS to match the version bump


## scReCkassify 0.99.4
* Minor change to example in `matPC` function


## scReCkassify 0.99.3
* Correct for any spelling errors in all documentations
* Revising the package to address the comments from Bioconductor review
* Cleaned the GSE87795 subset data to a SingleCellExperiment object


## scReCkassify 0.99.2
* version bump with submission to Bioconductor

## scReClassify 0.99.1
* Cleaning repository to pass BiocCheck

## scReClassify 0.99.0
* Initial submission for Bioconductor
* Code formats/documentations have been revised to meet Bioconductor 
requirements

## scReClassify 0.1.1

* Significant updates and addition to the documentations
* Major changes to code writing style to comply with BioC
* Updated example data


tests/testthat.R
library(testthat)
library(scReClassify)

test_check("scReClassify")


tests/testthat/test-matPCs.R
test_that(
    "Case matPCs.1: Missing parameters",
    {
        expect_error(matPCs())
        expect_error(matPCs(percentVar = 0.8))
    }
)

test_that(
    "Case matPCs.2: Expected outputs",
    {
        set.seed(123)
        # 50 cells 400 genes
        mat = matrix(rnorm(20000), nrow = 50)
        pc = prcomp(mat)
        eigs = pc$sdev^2
        p = cumsum(eigs)/sum(eigs)
        
        result1 = min(which(p > 0.5))
        result2 = min(which(p > 0.8))
        # Since result2 > 20
        result2 = 20L
        
        # Number of PCs returned
        expect_identical(result1, ncol(matPCs(t(mat), percentVar = 0.5)))
        expect_identical(result2, ncol(matPCs(t(mat), percentVar = 0.8)))
    }
)


tests/testthat/test-bAccuracy.R
test_that(
    "Case bAccuracy.1: Missing parameters",
    {
        cls = 1:12
        label = 1:12
        
        expect_error(bAccuracy())
        expect_error(bAccuracy(cls.truth = cls))
        expect_error(bAccuracy(final = label))
    }
)


test_that(
    "Case bAccuracy.2: Different parameter lengths",
    {
        cls = 1:12
        label = 1:10
        
        expect_error(bAccuracy(cls, label))
    }
)

test_that(
    "Case bAccuracy.3: Expected outputs",
    {
        cls = 1:12
        label = 12:1
        exp_result1 = 0
        exp_result2 = 1
        
        expect_identical(exp_result1, bAccuracy(cls, label))
        expect_identical(exp_result2, bAccuracy(cls, cls))
        
    }
)


R/data.R
#' @title GSE827795 subset data
#'
#' @description A SingleCellExperiment object containing a subset expression 
#' matrix of GSE827795. The data contains log2 transformed FPKM expression.
#' 
#' GSE87795 is a mouse fetal liver development data containing 1000 genes, 
#' 367 cells and 6 cell types. 
#' 
#' The original GSE87795 data and the study details can be found at this
#' [link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87795)
#'
#'
#' @name gse87795_subset_sce
#' @docType data
#'
"gse87795_subset_sce"



#' scReClassify: a package for  post hoc cell type classification of single-cell 
#' RNA-sequencing data.
#' 
#' @description A post hoc cell type classification tool to fine-tune cell type 
#' annotations generated by any cell type classification procedure with 
#' semi-supervised learning algorithm AdaSampling technique. 
#' 
#' The current version of scReClassify supports Support Vector Machine and 
#' Random Forest as a base classifier.
#'
#' @author 
#' **Maintainer**: 
#' 
#' * Taiyun Kim (ORCID:0000-0002-5028-836X)
#'     - Email: taiyun.kim91@gmail.com
#' 
#' Authors:
#' 
#' * Pengyi Yang (ORCID: 0000-0003-1098-3138)
#' 
#' 
#' @seealso 
#' 
#' Useful links:
#' * Vignette available at: https://sydneybiox.github.io/scdney/
#' 
#' @docType package
#' @name scReClassify
NULL



R/bAccuracy.R
#' bAccuracy
#'
#' This function calculates the accuracy of the prediction to the true label.
#'
#' @param cls.truth A character vector of true class label.
#' @param final A vector of final classified label prediction from 
#' \code{multiAdaSampling}.
#' @return An accuracy value.
#' @author Pengyi Yang, Taiyun Kim
#'
#' @examples
#' data("gse87795_subset_sce")
#'
#' mat.expr <- gse87795_subset_sce
#' cellTypes <- gse87795_subset_sce$cellTypes
#'
#' # Get dimension reduced matrix. We are using `logNorm` assay from `mat.expr`.
#' mat.pc <- matPCs(mat.expr, assay = "logNorm")
#'
#' # Here we are using Support Vector Machine as a base classifier.
#' result <- multiAdaSampling(mat.pc, cellTypes, classifier = "svm",
#' percent = 1, L = 10)
#'
#' final <- result$final
#'
#' # Balanced accuracy
#' bacc <- bAccuracy(cellTypes, final)
#'
#' @export bAccuracy
bAccuracy <- function(cls.truth, final) {
    
    if (length(cls.truth) != length(final)) {
        stop("`cls.truth` and `final` objects have different length.")
    }
    
    gs <- names(table(cls.truth))
    acc <- c()
    acc <- unlist(lapply(seq_len(length(gs)), function(i) {
        sum((cls.truth == final)[cls.truth==gs[i]]) / sum(cls.truth==gs[i])
    }))
    mean(acc)
}


R/multiAdaSampling.R
#' multi Adaptive Sampling function
#'
#' Performs multiple adaptive sampling to train a classifier model.
#'
#' @param data A dimension reduced matrix from \code{matPCs}.
#' @param label A vector of label information for each sample.
#' @param reducedDimName A name of the \code{reducedDim} to use. This must be 
#' specified if \code{data} is a SingleCellExperiment object.
#' @param classifier Base classifier model, either  "SVM" (\code{svm}) or "RF" 
#' \code{'rf'} is supported.
#' @param percent Percentage of samples to select at each iteration.
#' @param L Number of ensembles. Default to 10.
#' @param prob logical flag to return sample's probabilities to each class.
#' @param balance logical flag to if the cell types are balanced. 
#' If `FALSE`, down sample large cell types classes to the median of all class 
#' sizes.
#' @param iter A number of iterations to perform adaSampling.
#' @return A final prediction, probabilities for each cell type and the model
#' are returned as a list.
#' 
#' @author Pengyi Yang, Taiyun Kim
#' @importFrom randomForest randomForest
#' @importFrom e1071 svm
#' @importFrom stats median predict
#' @import SingleCellExperiment
#' @examples
#' 
#' library(SingleCellExperiment)
#' 
#' # Loading the data
#' data("gse87795_subset_sce")
#'
#' mat.expr <- gse87795_subset_sce
#' cellTypes <- gse87795_subset_sce$cellTypes
#' 
#' # Get dimension reduced matrix. We are using `logNorm` assay from `mat.expr`.
#' reducedDim(mat.expr, "matPCs") <- matPCs(mat.expr, assay = "logNorm")
#'
#' # Here we are using Support Vector Machine as a base classifier.
#' result <- multiAdaSampling(mat.expr, cellTypes, reducedDimName = "matPCs", 
#' classifier = "svm", percent = 1, L = 10)
#' @export multiAdaSampling
multiAdaSampling <- function(data, label, reducedDimName = NULL, 
                             classifier="svm", percent=1, L=10, prob=FALSE, 
                             balance=TRUE, iter=3) {
    # If SCE
    if (is(data, "SingleCellExperiment")) {
        if (is.null(reducedDimName)) {
            stop("reducedDimName parameter cannot be NULL if data is a", 
                 " SingleCellExperiment object")
        }
        mat <- SingleCellExperiment::reducedDim(data, reducedDimName)
    } else {
        mat <- data
    }
    mat <- t(mat)
    
    if (ncol(mat) != length(label)) {
        stop("Parameter `label` and number of columns of data must be equal")
    }
    if (percent < 0 | percent > 1) {
        stop("Parameter percent cannot must be a value between 0 and 1")
    }
    
    models <- lapply(seq_len(L), function(l) {
        X <- mat
        Y <- label
        
        model <- c()
        prob.mat <- c()
        
        for (i in seq_len(iter)) {
            if (classifier == "rf") {
                model <- randomForest::randomForest(t(X), factor(Y),
                                                    ntree = 100)
                prob.mat <- stats::predict(model, newdata=t(mat), type="prob")
            }
            if (classifier == "svm") {
                tmp <- t(X)
                rownames(tmp) <- NULL
                model <- e1071::svm(tmp, factor(Y), probability = TRUE)
                prob.mat <- attr(stats::predict(model, t(mat),
                                                decision.values = FALSE,
                                                probability = TRUE),
                                 "probabilities")
            }
            if (!(classifier %in% c("svm", "rf"))) {
                stop("Classifier ", classifier, " is unknown.")
            }
            
            X <- c()
            Y <- c()
            
            XY_update <- lapply(seq_len(ncol(prob.mat)), function(j) {
                voteClass <- prob.mat[label==colnames(prob.mat)[j],]
                idx <- c()
                if (balance) {
                    idx <- sample(seq_len(nrow(voteClass)),
                                  size=nrow(voteClass)*percent, replace = TRUE,
                                  prob=voteClass[,j])
                } else {
                    sampleSize <- round(stats::median(table(label)))
                    if (nrow(voteClass) > sampleSize) {
                        idx <- sample(seq_len(nrow(voteClass)), size=sampleSize*
                                          percent, replace = TRUE, 
                                      prob=voteClass[,j])
                    } else {
                        idx <- sample(seq_len(nrow(voteClass)),
                                      size=nrow(voteClass)*percent, 
                                      replace = TRUE, prob=voteClass[,j])
                    }
                }
                
                list(
                    X = mat[, rownames(voteClass)[idx]],
                    Y = label[rownames(voteClass)[idx]]
                )
            })
            cur_X = lapply(XY_update, function(xy) {
                xy$X
            })
            cur_Y = lapply(XY_update, function(xy) {
                xy$Y
            })
            X <- do.call(cbind, cur_X)
            Y <- do.call(c, cur_Y)
        }
        
        model
    })
    
    predictMat <- matrix(0, nrow=ncol(mat), ncol=length(table(label)))
    final <- c()
    predmat <- lapply(models, function(m) {
        if (classifier == "svm") {
            tmp <- attr(predict(m,
                                newdata=t(mat), probability = TRUE),
                        "prob")[,names(table(label))]
        } else {
            tmp <- predict(m, newdata=t(mat),
                           type="prob")[,names(table(label))]
        }
        predictMat <<- predictMat + tmp
    })
    
    if(prob==TRUE) {
        final <- apply(predictMat, 1, max)
        names(final) <- names(table(label))[apply(predictMat, 1, which.max)]
    } else {
        final <- names(table(label))[apply(predictMat, 1, which.max)]
    }
    
    return(
        list(
            final = final,
            models = models,
            prob = predictMat
        )
    )
}


R/predictAdaSample.R
#' @import stats
predictAdaSample <- function(models, data, labelNames, classifier = "svm",
                            prob = FALSE) {
    L = length(models)

    predictMat <- matrix(0, nrow=ncol(data), ncol=length(labelNames))
    final <- c()
    for (l in seq_len(L)) {
        if (classifier == "svm") {
            tmp <- attr(predict(models[[l]],newdata=t(data),probability = TRUE),
                        "prob")[,names(labelNames)]
            predictMat <- predictMat + tmp
        } else {
            tmp <- predict(models[[l]], newdata=t(data),
                            type="prob")[,names(labelNames)]
            predictMat <- predictMat + tmp
        }
    }
    if(prob==TRUE) {
        final <- apply(predictMat, 1, max)
        names(final) <- names(labelNames)[apply(predictMat, 1, which.max)]
    } else {
        final <- names(labelNames)[apply(predictMat, 1, which.max)]
    }

    return(final)
}


R/scReClassifyBench.R
# list.of.packages <- c("randomForest", "e1071", "gmodels", "mclust")
# new.package.list <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
# if (length(new.package.list)) install.packages(new.package.list)
#
# # potential packages
# # "amap", "aricode", "edgeR", "biomaRt"
#
# library(randomForest)
# library(e1071)
# library(gmodels)
#
#
# scReClassifyBench <- function(mat, label, seed, classifier="svm", L=10, prob=FALSE, percentVar=0.8) {
#   dat.selected <- matPCs(dat.processed, percentVar, seed = seed)
#
#   cellTypes <- label
#
#   colnames(dat.selected$pcs) <- paste("V", 1:ncol(dat.selected$pcs), sep = "")
#   names(cellTypes) <- paste("V", 1:ncol(dat.selected$pcs), sep = "")
#
#   pred <- multiAdaSampling(dat.selected$pcs, cellTypes, seed=seed, classifier=classifier, percent=1, L=L)$final
#   # rf.pred <- multiAdaSampling(dat.selected$pcs, cellTypes, seed=1, classifier="rf", percent=1, L=10)$final
#
#   adjustedRandIndex(pred, cellTypes)
#   # adjustedRandIndex(rf.pred, cellTypes)
#
#
#
# }


R/matPCs.R
#' matPCs function
#'
#' @description Performs PCA on a given matrix and returns a dimension reduced 
#' matrix which captures at least 80% (default) of overall variability.
#'
#' @param data An expression matrix or a SingleCellExperiment object.
#' @param assay An assay to select if \code{data} is a SingleCellExperiment 
#' object
#' @param percentVar The percentage of variance threshold. This is used to 
#' select number of Principal Components.
#' @details This function performs PCA to reduce the dimension of the gene
#' expression matrix limited from 10 to 20 PCs.
#' @return Dimensionally reduced matrix.
#' @author Pengyi Yang, Taiyun Kim
#' @export matPCs
#' @examples
#' data("gse87795_subset_sce")
#'
#' mat.expr <- gse87795_subset_sce
#'
#' mat.pc <- matPCs(mat.expr, assay = "logNorm")
#'
#' # to capture at least 70% of overall variability in the dataset,
#' mat.dim.reduct.70 <- matPCs(mat.expr, assay = "logNorm", 0.7)
#'
#' @importFrom stats prcomp sd
#' @importFrom SummarizedExperiment assay
#' @importFrom methods is
#' @import SingleCellExperiment
matPCs <- function(data, assay = NULL, percentVar=0.8) {
    
    if (percentVar < 0 | percentVar > 1) {
        stop("percentVar must be a number between 0 and 1")
    }
    # If SCE
    if (is(data, "SingleCellExperiment")) {
        if (is.null(assay)) {
            stop("assay parameter cannot be NULL")
        }
        mat <- SummarizedExperiment::assay(data, assay)
    } else {
        mat <- data
    }
    
    if (is(data, "matrix") | is(data, "data.frame")) {
        pca <- stats::prcomp(t(mat), center = TRUE, scale. = TRUE)
    } else if ((is(data, "SingleCellExperiment")) & 
        ("PCA" %in% reducedDimNames(data))) {
        pca <- SingleCellExperiment::reducedDim(data, "PCA")
    } else if (is(data, "SingleCellExperiment")) {
        pca <- stats::prcomp(t(mat), center = TRUE, scale. = TRUE)
        if (length(SingleCellExperiment::reducedDimNames(data))) {
            rd <- SingleCellExperiment::reducedDims(data)
            rd$PCA = pca$x
        } else {
            SingleCellExperiment::reducedDims(data) <- list(PCA = pca$x)
        }
    }

    # genes are rows, cells are cols
    pcs <- c()
    sdev <- apply(pca$x, 2, sd)
    eigs <- sdev^2
    top <- which(cumsum(eigs)/sum(eigs) > percentVar)[1]
    m <- ifelse(top > 20, 20, top)
    m <- ifelse(m < 10, 10, m)
    pcs <- pca$x[,seq_len(m)]

    return(pcs)
}
